packages/driver/src/httpScram.ts

```diff
@@ -41,27 +66,42 @@ export async function HTTPSCRAMAuth(
       Authorization: `SCRAM-SHA-256 data=${utf8ToB64(clientFirst)}`,
     },
   });
-  if (serverFirstRes.status === 403) {
-    // tslint:disable-next-line:no-console
-    console.log(serverFirstRes);
-    throw new Error(`Server doesn't support HTTP SCRAM authentication`);
+
+  // The first request must have status 401 Unauthorized and provide a
+  // WWW-Authenticate header with a SCRAM-SHA-256 challenge.
+  // See: https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L153-L157
+  const authenticateHeader = serverFirstRes.headers.get("WWW-Authenticate");
+  if (serverFirstRes.status !== 401 || !authenticateHeader) {
+    const body = await serverFirstRes.text();
+    throw new ProtocolError(`authentication failed: ${body}`);
   }
-  const firstAttrs = parseHeaders(serverFirstRes.headers, "WWW-Authenticate");
-  if (firstAttrs.size === 0) {
-    throw new Error("Invalid credentials");
+
+  // WWW-Authenticate can contain multiple comma-separated authentication
+  // schemes (each with own comma-separated parameter pairs), but we only support
+  // one SCRAM-SHA-256 challenge, e.g., `SCRAM-SHA-256 sid=..., data=...`.
+  if (!authenticateHeader.startsWith("SCRAM-SHA-256")) {
+    throw new ProtocolError(
+      `unsupported authentication scheme: ${authenticateHeader}`
+    );
   }
-  if (!firstAttrs.has("sid") || !firstAttrs.has("data")) {
+
+  // The server may respond with a 401 Unauthorized and `WWW-Authenticate: SCRAM-SHA-256` with
+  // no parameters if authentication fails, e.g., due to an incorrect username.
+  // See: https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L112-L120
+  const authParams = authenticateHeader.split(/ (.+)?/, 2)[1] ?? "";
+  if (authParams.length === 0) {
+    const body = await serverFirstRes.text();
+    throw new ProtocolError(`authentication failed: ${body}`);
```

Comment:
I think in this case adding the `authenticateHeader` text is probably the most helpful thing here instead of (or in addition to?) the body text. Maybe also log the header keys?

Reply:
I believe the only value `authenticateHeader` could have here is `SCRAM-SHA-256` (for values starting with anything but `SCRAM-SHA-256` we would've thrown on [line 84](https://github.com/edgedb/edgedb-js/pull/620/files/d3ade318144f83e73d010a83f2c3cbe9f9ee9ec0#diff-58b536d06e962df6fbe8c6ddaffd1a334bd0001566bc4a5f606b11e99e2dba2cR84), and the rest of the header was non-empty, we wouldn't be in this branch). However, the error message may either be [`"Client selected an invalid SASL authentication mechanism"`](https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L39-L45) or [`"Authentication failed"`](https://github.com/edgedb/edgedb/blob/09782afd3b759440abbb1b26ee19b6589be04275/edb/server/protocol/auth/scram.py#L115-L120) here.